### PR TITLE
feat!: Align module with latest Snowflake provider changes (removals and deprecations

### PR DESCRIPTION
**Pull request by dpnoworyta**
This PR implements **BREAKING CHANGES!!!**

This PR completely rewrites the module, to align it with Snowflake provider latest versions (`>0.94`), implements some code optimizations and changes API (variables) in significant way.

List of code and variable (API) changes:
- Switched to `snowflake_database_role` module to leverage new `snowflake_database_roles` mechanism
- stage module version was updated (`>= v2.1.0`: https://github.com/getindata/terraform-snowflake-stage/pull/16) + https://github.com/getindata/terraform-snowflake-stage/pull/18 to use newly introduced changes by Snowflake provider
- overhaul of `roles` and `stages` variables
- variable `add_grants_to_existing_objects` was removed as it is no longer needed

### CODE REVIEW

File: locals.tf [hunk 1]
```diff
@@ -5,86 +5,420 @@ locals {
...
+    admin = {
+      schema_grants = [{
+        privileges              = ["ALL PRIVILEGES"]
+        all_schemas_in_database = true
+      }]
+      schema_objects_grants = {
+        for object_type in ["TABLE", "DYNAMIC TABLE", "EXTERNAL TABLE", "VIEW", "MATERIALIZED VIEW", "FILE FORMAT", "FUNCTION", "STAGE", "TASK", "PROCEDURE"] :
+        object_type => [{
+          privileges = ["ALL PRIVILEGES"]
+          on_all     = true
```
Comment (dpnoworyta) [inline locals.tf:46]: should default roles have
`on_future = true`
enabled by default for all objects_type that they have?

File: locals.tf [hunk 2]
```diff
@@ -5,86 +5,420 @@ locals {
...
+            tostring([])
+          ]
+          on_all = true
+        }]
+      }
     }
-  }
+  } : {}
+
+  # default_roles_definition = var.create_default_database_roles ? {
```
Comment (dpnoworyta) [inline locals.tf:134]: I've created two `default_roles_definition` - which should we use?

File: locals.tf (hunk 1 above)
Comment (moss2k13) [reply locals.tf:46]: i would say yes, it sounds like a reasonable default setting

Comment (moss2k13) [reply locals.tf:46]: we still are able to override `on_future = true` default value if needed right?

File: locals.tf [hunk 3]
```diff
@@ -5,86 +5,420 @@ locals {
...
-      materialized_view_grants = ["SELECT", "REFERENCES"]
-      file_format_grants       = ["USAGE"]
-      function_grants          = ["USAGE"]
-      stage_grants             = ["USAGE"]
-      task_grants              = ["MONITOR"]
+  default_roles_definition = var.create_default_database_roles ? {
+    admin = {
+      schema_grants = [{
+        privileges              = ["ALL PRIVILEGES"]
+        all_schemas_in_database = true
```
[the remark this answers is not in the document]
Comment (moss2k13) [reply locals.tf:40]: i'd leave the current value `future_schemas_in_database = true`

File: locals.tf [hunk 4]
```diff
@@ -5,86 +5,420 @@ locals {
...
+        for object_type in ["TABLE", "DYNAMIC TABLE", "EXTERNAL TABLE", "VIEW", "MATERIALIZED VIEW", "FILE FORMAT", "FUNCTION", "STAGE", "TASK"] :
+        object_type => [{
+          privileges = [
+            object_type == "TABLE" || object_type == "DYNAMIC TABLE" ? join(",\n", ["SELECT"]) :
+            object_type == "EXTERNAL TABLE" || object_type == "VIEW" || object_type == "MATERIALIZED VIEW" ? join(",\n", ["SELECT", "REFERENCES"]) :
+            object_type == "FILE FORMAT" || object_type == "FUNCTION" || object_type == "STAGE" ? join(",\n", ["USAGE"]) :
+            object_type == "TASK" ? join(",\n", ["MONITOR"]) :
+            tostring([])
+          ]
+          on_all = true
```
Comment (moss2k13) [inline locals.tf:88]: will we be able to override default value for `on_all = true` for specific object_type while using this compact `for` loop?

Comment (dpnoworyta) [reply locals.tf:88]: yes, eg:
```
+ transformer     = {
[...]
  + "EXTERNAL TABLE"    = [
      + {
          + all_privileges    = true
          + object_name       = null
          + on_all            = true
          + on_future         = true
          + privileges        = null
          + schema_name       = null
          + with_grant_option = true
        },
    ]
  + "FILE FORMAT"       = [
      + {
          + all_privileges    = true
          + object_name       = null
          + on_all            = false
          + on_future         = true
          + privileges        = null
          + schema_name       = null
          + with_grant_option = false
        },
    ]
```

File: locals.tf [hunk 5]
```diff
@@ -5,86 +5,428 @@ locals {
...
-      view_grants              = ["SELECT", "REFERENCES"]
-      materialized_view_grants = ["SELECT", "REFERENCES"]
-      file_format_grants       = ["USAGE"]
-      function_grants          = ["USAGE"]
-      stage_grants             = ["USAGE"]
-      task_grants              = ["MONITOR"]
+  default_roles_definition = var.create_default_database_roles ? {
+    admin = {
+      schema_grants = [{
+        privileges                 = ["ALL PRIVILEGES"]
```
Comment (dgniewek) [inline locals.tf:39]: We should use this flag `all_privileges = true` instead of `privileges = ["ALL PRIVILEGES"]`

File: locals.tf [hunk 6]
```diff
@@ -5,86 +5,428 @@ locals {
...
+        all_schemas_in_database    = true
+        future_schemas_in_database = true
+      }]
+      schema_objects_grants = {
+        for object_type in ["TABLE", "DYNAMIC TABLE", "EXTERNAL TABLE", "VIEW", "MATERIALIZED VIEW", "FILE FORMAT", "FUNCTION", "STAGE", "TASK", "PROCEDURE"] :
+        object_type => [{
+          privileges = ["ALL PRIVILEGES"]
```
Comment (dgniewek) [inline locals.tf:46]: We should use this flag `all_privileges = true` instead of `privileges = ["ALL PRIVILEGES"]`

File: variables.tf [hunk 7]
```diff
@@ -33,70 +43,57 @@ variable "is_managed" {
...
-variable "add_grants_to_existing_objects" {
-  description = "Whether grants to existing objects should be added"
+variable "create_default_database_roles" {
+  description = "Whether the default database roles should be created"
   type        = bool
   default     = false
 }
 
 variable "roles" {
   description = "Roles created in the scheme scope"
```
Comment (dgniewek) [inline variables.tf:53]: Please change description - "Database roles..."

File: variables.tf [hunk 8]
```diff
@@ -33,70 +43,57 @@ variable "is_managed" {
...
-    comment              = optional(string)
-    copy_options         = optional(string)
-    credentials          = optional(string)
-    directory            = optional(string)
-    encryption           = optional(string)
-    file_format          = optional(string)
-    snowflake_iam_user   = optional(string)
-    storage_integration  = optional(string)
-    url                  = optional(string)
-    create_default_roles = optional(bool)
```
Comment (dgniewek) [inline variables.tf:90]: Why this flag is deleted?

File: variables.tf [hunk 9]
```diff
@@ -33,70 +43,57 @@ variable "is_managed" {
...
-    roles = optional(map(object({
```
Comment (dgniewek) [inline variables.tf:91]: Why role definitions are deleted?

File: variables.tf [hunk 10]
```diff
@@ -33,70 +43,57 @@ variable "is_managed" {
...
-    task_grants                    = optional(list(string))
-    procedure_grants               = optional(list(string))
-    sequence_grants                = optional(list(string))
-    stream_grants                  = optional(list(string))
+    enabled                   = optional(bool, true)
+    role_ownership_grant      = optional(string)
+    granted_to_roles          = optional(list(string))
+    granted_to_database_roles = optional(list(string))
+    granted_database_roles    = optional(list(string))
+    database_grants           = optional(list(string))
```
Comment (moss2k13) [inline variables.tf:60]: based on:
https://github.com/getindata/terraform-snowflake-database-role/blob/96bf8f6a78443489a39f7b8883dba3c0ff9f06d2/variables.tf#L41-L48
```suggestion
    database_grants           = optional(object({
      all_privileges    = optional(bool)
      with_grant_option = optional(bool, false)
      privileges        = optional(list(string), null)
    }))
```

File: examples/complete/main.tf [hunk 11]
```diff
@@ -1,54 +1,126 @@
...
 
   stages = {
-    my_stage = {}
+    my_stage = {
+      comment = "Stage used to ingest data"
+    }
   }
 
-  create_default_roles = true
+  create_default_database_roles = true
```
[the remark this answers is not in the document]
Comment (dpnoworyta) [reply examples/complete/main.tf:69]: I did it similarly to this one: https://github.com/getindata/terraform-snowflake-stage/blob/main/variables.tf#L77

File: locals.tf (hunk 5 above)
Comment (dpnoworyta) [reply locals.tf]: Setting this flag breaks conditional `default_role_definition` creation - I would like to stick to this formulae. 
Looking at provider files, `["ALL PRIVILEGES"]` are respected: https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/templates/resources/grant_privileges_to_database_role.md.tmpl#L41

File: examples/complete/main.tf (hunk 11 above)
[the remark this answers is not in the document]
Comment (dgniewek) [reply examples/complete/main.tf:69]: Please rollback that flag's name in snowflake-stage also - we're trying to maintain an similar API i every module

File: locals.tf (hunk 5 above)
Comment (dgniewek) [reply locals.tf]: Since that works fine - we can stick with your solution

File: main.tf [hunk 12]
```diff
@@ -44,360 +61,45 @@ module "snowflake_stage" {
...
+  version = "1.1.1"
   context = module.this.context
-  enabled = local.create_default_roles && each.value.enabled
+  enabled = local.create_default_roles && lookup(each.value, "enabled", true)
 
-  name       = each.key
-  attributes = [local.database, local.schema]
+  database_name = one(snowflake_schema.this[*].database)
+  name          = each.key
+  attributes    = [local.database, local.schema]
```
Comment (dgniewek) [inline main.tf:79]: We don't need DB name in DB role attributes

```suggestion
  attributes    = [local.schema]
```

File: main.tf [hunk 13]
```diff
@@ -44,360 +61,45 @@ module "snowflake_stage" {
...
-  privilege     = each.key
-  roles         = each.value
-}
+  enabled = module.this.enabled && lookup(each.value, "enabled", true)
 
-resource "snowflake_stream_grant" "existing" {
-  for_each = module.this.enabled ? local.stream_grants_on_existing : {}
+  database_name = one(snowflake_schema.this[*].database)
+  name          = each.key
+  attributes    = [local.database, local.schema]
```
Comment (dgniewek) [inline main.tf:98]: The same as above

```suggestion
  attributes    = [local.schema]
```

File: outputs.tf [hunk 14]
```diff
@@ -28,7 +13,12 @@ output "stages" {
...
 }
 
-output "roles" {
-  description = "Snowflake Roles"
+output "database_roles" {
+  description = "Snowflake Database Roles"
   value       = local.roles
 }
+
+output "schema" {
```
Comment (dgniewek) [inline outputs.tf:21]: Please split this into separate outputs

**Review by dgniewek — CHANGES_REQUESTED** [review]

File: variables.tf [hunk 15]
```diff
@@ -15,88 +18,99 @@ variable "comment" {
...
-    materialized_view_grants       = optional(list(string))
-    file_format_grants             = optional(list(string))
-    function_grants                = optional(list(string))
-    stage_grants                   = optional(list(string))
-    task_grants                    = optional(list(string))
-    procedure_grants               = optional(list(string))
-    sequence_grants                = optional(list(string))
-    stream_grants                  = optional(list(string))
+    enabled                   = optional(bool, true)
+    descriptor_name           = optional(string, "snowflake-role")
```
Comment (dgniewek) [inline variables.tf:56]: `snowflake-database-role`

File: variables.tf [hunk 16]
```diff
@@ -15,88 +18,99 @@ variable "comment" {
...
+    comment             = optional(string)
+    copy_options        = optional(string)
+    credentials         = optional(string)
+    directory           = optional(string)
+    encryption          = optional(string)
+    file_format         = optional(string)
+    snowflake_iam_user  = optional(string)
+    storage_integration = optional(string)
+    url                 = optional(string)
     roles = optional(map(object({
```
Comment (dgniewek) [inline variables.tf:102]: `descriptor_name` is missing here

File: main.tf [hunk 17]
```diff
@@ -44,360 +61,47 @@ module "snowflake_stage" {
...
   context = module.this.context
-  enabled = local.create_default_roles && each.value.enabled
+  enabled = local.create_default_roles && lookup(each.value, "enabled", true)
 
-  name       = each.key
-  attributes = [local.database, local.schema]
+  database_name   = one(snowflake_schema.this[*].database)
+  name            = each.key
+  attributes      = [local.schema]
+  descriptor_name = lookup(each.value, "descriptor_name", "snowflake-role")
```
Comment (dgniewek) [inline main.tf:80]: `snowflake-database-role`

File: main.tf [hunk 18]
```diff
@@ -44,360 +61,47 @@ module "snowflake_stage" {
...
-
-  database_name = local.database
-  schema_name   = local.schema
-  on_all        = true
-  privilege     = each.key
-  roles         = each.value
+  database_name   = one(snowflake_schema.this[*].database)
+  name            = each.key
+  attributes      = [local.schema]
+  descriptor_name = lookup(each.value, "descriptor_name", "snowflake-role")
```
Comment (dgniewek) [inline main.tf:100]: `snowflake-database-role`

**Review by dgniewek — CHANGES_REQUESTED** [review]

**Review by dgniewek — APPROVED** [review]

**Pull request merged**